### PR TITLE
Create new structured facts for `gluster_peers` and `gluster_volumes`

### LINKED ISSUE #33: Weak regex for volume port Fact
The regex used to feed the volume port Fact can't always parse completely the output of the 'gluster volume status' command.
[gluster.rb](https://github.com/voxpupuli/puppet-gluster/blob/master/lib/facter/gluster.rb) L 48

``` ruby
volume_ports[vol] = status.scan(/^Brick [^\t]+\t+(\d+)/).flatten.uniq.sort
```

I suspect that it might happen when the brick name is long enough to be formatted into two lines.

The end result is an empty Fact in Puppet

### LINKED ISSUE #53: Support commas in volume options
Currently, volume options are collected as comma seperated list using the `gluster_volume_${title}_options` fact. However, some volume options may have values containing commas as well (for example, `nfs.rpc-auth-allow`).

In order to make this work, I see two approaches:
- Use another seperating character in the list of volume options provided by the `gluster_volume_${title}_options` fact (this should ideally be a character, which would never ever occur in volume option names and values). For example, `gluster_volume_${title}_options` could be something like `option1: 'value1', option2: 'value2,value3'`
- Use Facter arrays/hashes (which are, AFAIK, guaranteed to be available only for Puppet >= 4.0)

### LINKED ISSUE #165: Volume mapping values for auth.allow with multiple ips
Hi,

I keep getting the error below when I try to set multiple ips for "auth.allow" for a volume and I can't seem to figure out why or what I might be doing wrong.

```
Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, (<unknown>): mapping values are not allowed here at line 2 column 25 at /etc/puppetlabs/code/environments/test/modules/gluster/manifests/volume.pp:276:21 at /etc/puppetlabs/code/environments/test/manifests/classes/test/gluster/glusterd_volume.pp:52 on node example.com
Warning: Not using cache on failed catalog
Error: Could not retrieve catalog; skipping run
```

Could someone help me figureout what I might be doing wrong?

### PR DESCRIPTION
**Pull request by tparkercbn**
#### Created new structured facts for gluster_peers and gluster_volumes using the --xml cli modifier.

All legacy facts are unchanged to not cause any breaking changes to people who may be using them

#### Pull Request (PR) description
Created new structured facts for gluster_peers and gluster_volumes that fix the problems caused by comma separated data structures containing comma separated values.  It also fixes the gluster_volume_ports bug when the volume names line wrap

#### This Pull Request (PR) fixes the following issues
Fixes #165 
Fixes #53
Fixes #33

### CODE REVIEW

**Comment by tparkercbn** [comment]
This pull request replaces my previous one that ended up being a complete mess of commits.  I'm apparently not great with Git :)

**Comment by bastelfreak** [comment]
Hi @tparkercbn, thanks for the PR. Can you please add a test test for the changes?

**Comment by tparkercbn** [comment]
I have no idea how to do that :)  Let me see what I can figure out.

**Comment by tparkercbn** [comment]
Are there any current tests for the output of the facts?  My code makes no changes at all to the functionality of the module itself other than fixing the comma separated values bug in the volume options

**Comment by alexjfisher** [comment]
@tparkercbn Would some examples help?

https://github.com/voxpupuli/puppet-nginx/blob/master/spec/unit/facter/util/fact_nginx_version_spec.rb
https://github.com/voxpupuli/puppet-mongodb/blob/master/spec/unit/mongodb_version_spec.rb

If you're really stuck, let us know.

**Comment by tparkercbn** [comment]
Thanks!  That should help a ton!

**Comment by bastelfreak** [comment]
Docker + systemd + Xorg are sometimes complicated. It is totally fine to just push changes and watch the outcome on travis. It will execute the tests for you. You just need to squash the git history afterwards.

**Comment by tparkercbn** [comment]
Ok cool.  I was hoping to be able to learn a bit more about the tests before making my code public :D but that's what I will do.

Thanks!

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

File: spec/unit/facter/util/fact_gluster_peers_spec.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+              <connected>1</connected>
+              <state>3</state>
+              <stateStr>Peer in Cluster</stateStr>
+            </peer>
+          </peerStatus>
+        </cliOutput>'
+      }
+    end
+
+    gluster_peers = {}
```
Comment (ekohl) [inline spec/unit/facter/util/fact_gluster_peers_spec.rb]: You can specify hashes inline:
```ruby
{
  "peer2" => {
    "uuid" => "7d1148a2-f19e-4f18-818f-3396ddf38c30",
      "connected" => 1,
      "state" => 3,
      "status" => "Peer in Cluster",
  },
  "peer3" => {
    "uuid" => "35f53c52-83dc-4100-a1f7-4a7cdeee074d",
    "connected" => 1,
    "state" => 3,
    "status" => "Peer in Cluster",
  },
}
```

Possibly Rubocop has something about this, but the idea is clear I hope ;)

**Comment by tparkercbn** [comment]
Sorry. I haven't forgotten about this but work and travel has had me up to my eyeballs.  

For some reason it's not finding the data that I am passing to validate.  Is there someone who can take a look and make some suggestions?

Thanks!

Tom

**Comment by tparkercbn** [comment]
This is awesome!  Thanks 😊

I applied the patches and pushed the change but we are failing the rubocop checks in travis-ci

I’m not sure how to properly give you credit for this work.  Maybe someone else on the team knows

Tom

**Comment by tryfunc** [comment]
Cool! Thank!
One caveat, the **gluster.rb.patch.txt** and **gluster_spec.rb.patch.txt** patch files are not needed in the repository. It would be better to remove them.

Oh, one test failed. Not quite the correct syntax. Now I will fix it.

**Comment by tparkercbn** [comment]
Yeah,  I just stashed them there and then forgot to remove before commit.  Still on my first coffee 😊

I was going to try to cleanup the indenting but you are probably faster 😊

Tom

**Comment by tryfunc** [comment]
Attached is the revised version of gluster _spec.rb. Checked several times - everything is ok.

[gluster_spec.rb.txt](https://github.com/voxpupuli/puppet-gluster/files/3407901/gluster_spec.rb.txt)

**Comment by tparkercbn** [comment]
Thank you.

All tests look good!

Now I will have to figure out how to rebase this all on master so it’s not a mess of commits 😊

I really appreciate your help on this.  I am still very new to rspec testing

Tom

**Comment by bastelfreak** [comment]
Hey @tparkercbn. If you need some help with rebasing you can ping us in our IRC channel #voxpupuli on freenode or on https://slack.puppet.com/

**Pull request merged**